network/benchmark/src/lib.rs

```diff
@@ -273,57 +253,175 @@ pub async fn run_netbench_service(
         },
     };
     let (work_sender, work_receiver) = async_channel::bounded(num_threads * 2);
-    let runtime_handle = Handle::current();
+    let listener_task = runtime_handle.spawn(connection_listener(
+        node_config.clone(),
+        network_client.clone(),
+        time_service.clone(),
+        shared.clone(),
+        runtime_handle.clone(),
+    ));
     let source_task = runtime_handle.spawn(source_loop(network_requests, work_sender));
     let mut handlers = vec![];
+    info!("netbench starting {} threads", num_threads);
     for _ in 0..num_threads {
         handlers.push(runtime_handle.spawn(handler_task(
-            node_config.clone(),
             network_client.clone(),
             work_receiver.clone(),
             time_service.clone(),
             shared.clone(),
         )));
     }
-    if let Err(err) = source_task.await {
-        warn!("benchmark source_thread join: {}", err);
-    }
+    let listener_task_result = listener_task.await;
+    info!("netbench listener_task exited {:?}",listener_task_result);
+    let source_task_result = source_task.await;
+    info!("netbench source_task exited {:?}", source_task_result);
     for hai in handlers {
         if let Err(err) = hai.await {
-            warn!("benchmark handler_thread join: {}", err);
+            warn!("netbench handler_thread[] join: {}", err);
+        }
+    }
+}
+
+async fn connection_listener(
+    node_config: NodeConfig,
+    network_client: NetworkClient<NetbenchMessage>,
+    time_service: TimeService,
+    shared: Arc<RwLock<NetbenchSharedState>>,
+    handle: Handle,
+) {
+    let config = node_config.netbench.unwrap();
+    if let Ok(peers) = network_client.peers_and_metadata.get_connected_peers_and_metadata() {
+        info!("netbench connection_listener got {} initial peers", peers.len());
+        for peer_network_id in peers.keys() {
+            info!("netbench connection_listener new for initial {:?}", peer_network_id);
+            if config.enable_direct_send_testing {
+                handle.spawn(direct_sender(
+                    node_config.clone(),
+                    network_client.clone(),
+                    time_service.clone(),
+                    peer_network_id.network_id(),
+                    peer_network_id.peer_id(),
+                    shared.clone(),
+                    handle.clone(),
+                ));
+            }
+            if config.enable_rpc_testing {
+                handle.spawn(rpc_sender(
+                    node_config.clone(),
+                    network_client.clone(),
+                    time_service.clone(),
+                    peer_network_id.network_id(),
+                    peer_network_id.peer_id(),
+                    shared.clone(),
+                ));
+            }
+        }
+    }
+    let mut connection_notifications = network_client.peers_and_metadata.subscribe();
```

Comment:
Interesting... I guess a subscription model means that we don't unnecessarily need to keep looping over the existing set (to identify new and lost peers)?